### PR TITLE
routeDiscoveryDone Type and Order Posts

### LINKED ISSUE #13: RSS: Add Option to Reverse Items in the Feed
In order to use the feed to automate sending emails, I need the newest posts first, but this only allows for latest posts. I'm thinking an option in the config like: 

```
newestPostsFirst: true | false
```

### PR DESCRIPTION
**Pull request by pjlamb12**
feat: convert to routeDiscoveryDone

Convert the plugin to use routeDiscoveryDone instead of render so it runs once instead of once per route
add the option to order by newest first or oldest first

Scully doesn't seem to pass the dates for the route data correctly. The dates are an empty object. I have asked on Gitter/Discord for clarification on this.

Breaking Changes Plugin won't be run as a postRenderer for a route
Closes #13 and #14

### CODE REVIEW

**Comment by pjlamb12** [comment]
I'm working on getting the tests fixed on this.

**Comment by pjlamb12** [comment]
I think the last error is unrelated to the changes I've made, but I'm trying to hear back from the Scully Team.

**Comment by pjlamb12** [comment]
I figured out what the last error is. The last error is that when the RSS feed is to be written, the `date.toUTCString` function fails because the date fields on the `HandledRoute` are empty objects for some reason. I'm not sure why they're being returned as empty objects, currently.

**Comment by marcjulian** [comment]
@pjlamb12 thanks for looking into improving the RSS feed plugin!! I tried it out locally as well and for some reason the `publishedAt` and `updatedAt` dates are empty. I thought it might be because scully deps was install as the beta version. After updating it to latest it still fails, updates are added to master. 

Here is an example for another `routeDiscoveryDone` plugin, maybe this helps a bit https://github.com/gammastream/scully-plugins/tree/master/projects/scully-plugin-sitemap

**Comment by marcjulian** [comment]
Changing the plugin to `allDone` has the same result as `publishedAt` and `updatedAt` are empty objects.

```
{
  postRenderers: [ 'fouc', 'seoHrefOptimise', 'lazyImages', 'mediumZoom' ],
  route: '/blog/2020-03-21-food',
  type: 'contentFolder',
  templateFile: '/Users/m/Dev/notiz/scully-plugins/scully-plugin-lazy-images/content/blog/2020-03-21-food.md',
  data: {
    name: undefined,
    title: 'Food',
    description: 'Food description',
    publishedAt: {},
    updatedAt: {},
    published: true,
    tags: [ 'lazy-images' ],
    authors: [ 'Bruce Lee' ],
    sourceFile: '2020-03-21-food.md'
  },
  config: {
    type: 'contentFolder',
    slug: { folder: './content/blog' },
    postRenderers: [ 'fouc', 'seoHrefOptimise', 'lazyImages', 'mediumZoom' ]
  }
}
```

**Comment by marcjulian** [comment]
Changing the format of the date in the frontmatter like

```md
---
title: Food
description: Food description
publishedAt: 2020-03-25T
updatedAt: 2020-03-25T
published: true
tags: [lazy-images]
authors:
  - Bruce Lee
---

# Food
```

It is correctly available in the plugin

```
{
  postRenderers: [ 'fouc', 'seoHrefOptimise', 'lazyImages', 'mediumZoom' ],
  route: '/blog/2020-03-21-food',
  type: 'contentFolder',
  templateFile: '/Users/m/Dev/notiz/scully-plugins/scully-plugin-lazy-images/content/blog/2020-03-21-food.md',
  data: {
    name: undefined,
    title: 'Food',
    description: 'Food description',
    publishedAt: '2020-03-25T',
    updatedAt: '2020-03-25T',
    published: true,
    tags: [ 'lazy-images' ],
    authors: [ 'Bruce Lee' ],
    sourceFile: '2020-03-21-food.md'
  },
  config: {
    type: 'contentFolder',
    slug: { folder: './content/blog' },
    postRenderers: [ 'fouc', 'seoHrefOptimise', 'lazyImages', 'mediumZoom' ]
  }
}
```

However, now it is not an valid date string.

**Comment by pjlamb12** [comment]
@marcjulian yeah, but if you change the dates to strings, just by wrapping them in quotes, then the string value comes through to the plugin. But the RSS Feed npm package then fails because it can't call a function on the string.

**Comment by marcjulian** [comment]
Okay as string it is correctly parsed to the plugin. That leaves us to the question why is the date not correctly available during the plugin process if it is not a string. I saw your discussion in Scully Gitter.

What if the date is a string in the front matter and the plugin transforms that string into a date for the RSS Feed package?

**Comment by pjlamb12** [comment]
I thought about that, but it felt like a bad workaround. The dates shouldn't be empty objects and finding the solution there felt better. But I'm open to whatever. If we can't figure out what's going on in scully, I guess we'd have to do that.

**Comment by marcjulian** [comment]
Yes building a workaround doesn't feel right, would be better if scully doesn't parse the dates as empty objects. Let's see if we can find something out with scully. If that doesn't work out we could start over with the workaround. What do you think would be the best approach?

**Comment by pjlamb12** [comment]
I think the best workaround approach would be having those dates be strings, like:

`publishedAt: '2020-01-01'`

And then in the `createFeedItemFromRoute` function in the rss.js file, we could convert the date from a string to a date:

`date: route.data.updatedAt ? new Date(route.data.updatedAt) : new Date(route.data.publishedAt)`

What do you think?

**Comment by marcjulian** [comment]
Yes sounds good. What if we want to keep the time as well? 

`publishedAt: '2020-03-25T10:12:00.000Z'`

And then we use for example [dayjs](https://day.js.org/) to convert it to a date

`date: route.data.updatedAt ? dayjs(route.data.updatedAt) : dayjs(route.data.publishedAt)`

**Comment by pjlamb12** [comment]
That would be great too. Works for me.

**Comment by marcjulian** [comment]
I just tried it out and it would work with dayjs like so

`date: route.data.updatedAt ? dayjs(route.data.updatedAt).toDate() : dayjs(route.data.publishedAt).toDate()`

**Comment by pjlamb12** [comment]
what happens if `route.data.updatedAt` or `route.data.publishedAt` is already a date object? Because if it works, then we could just go this route and if/when Scully passes the date as an actual date object instead of an empty object, then the plugin would still work.

**Comment by marcjulian** [comment]
I think dayjs would be fine either way with the string or the actual date object.

**Comment by pjlamb12** [comment]
Maybe we should just go ahead and add dayjs and the plugin would work either way and the release could be done.

**Comment by marcjulian** [comment]
I tried it out on the `master` branch. The routes value is as following

```
{
  postRenderers: [ 'fouc', 'seoHrefOptimise', 'lazyImages', 'mediumZoom', 'rss' ],
  route: '/blog/2020-03-21-food',
  type: 'contentFolder',
  templateFile: '/Users/m/Dev/notiz/scully-plugins/scully-plugin-lazy-images/content/blog/2020-03-21-food.md',
  data: {
    name: undefined,
    title: 'Food',
    description: 'Food description',
    publishedAt: 2020-03-25T10:12:00.000Z,
    updatedAt: 2020-03-25T10:12:00.000Z,
    published: true,
    tags: [ 'lazy-images' ],
    authors: [ 'Bruce Lee' ],
    sourceFile: '2020-03-21-food.md'
  },
  config: {
    type: 'contentFolder',
    slug: { folder: './content/blog' },
    postRenderers: [ 'fouc', 'seoHrefOptimise', 'lazyImages', 'mediumZoom', 'rss' ]
  }
}
```

I installed dayjs and wrapped the date as described above and it workes fine.

**Comment by marcjulian** [comment]
Do you want to add dayjs to the plugin and push again?

**Comment by pjlamb12** [comment]
Yes I will do that.

**Comment by marcjulian** [comment]
Thanks for looking into this 💯 much appreciated!

**Comment by pjlamb12** [comment]
How do I import dayjs into the plugin's rss.js file? I npm installed it, and required dayjs at the top but that isn't working.

**Comment by marcjulian** [comment]
const dayjs = require("dayjs");

**Comment by marcjulian** [comment]
Do you mind adding `dayjs` to the rss plugin and not in the root package.json?

File: plugins/rss/package.json [hunk 1]
```diff
@@ -33,6 +37,7 @@
   },
   "dependencies": {
     "feed": "^4.1.0",
-    "showdown": "^1.9.1"
+    "showdown": "^1.9.1",
+    "dayjs": "^1.9.6"
```
Comment (marcjulian) [inline plugins/rss/package.json:41]: Yes that's correct here!

File: package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "@scullyio/init": "1.0.0-beta.0",
     "@scullyio/ng-lib": "1.0.0-beta.0",
     "@scullyio/scully": "1.0.0-beta.0",
+    "dayjs": "^1.9.6",
```
Comment (marcjulian) [inline package.json:32]: Not necessary here.
```suggestion
```

**Comment by pjlamb12** [comment]
👍🏾 I did add it to the plugin but it was also in the root. I have pulled it out of the root.

**Review by marcjulian — CHANGES_REQUESTED** [review]

**Comment by pjlamb12** [comment]
Do you know why it's saying: "Cannot find module webpack"?

**Comment by marcjulian** [comment]
Yes might be because of tailwind which is added to the root dependencies. Let me have a look, sorry for the inconvenience

**Comment by marcjulian** [comment]
Works now. What about adding some logs for the user to the plugin? How about

```bash
Started @notiz/scully-plugin-rss
Generating RSS Feed for 2 blog posts
✅ Created ./dist/static/feed.xml
✅ Created ./dist/static/feed.atom
✅ Created ./dist/static/feed.atom
Finished @notiz/scully-plugin-rss
```

Screenshot 2020-11-23 at 19 23 47

**Review by marcjulian — APPROVED** [review]

**Comment by pjlamb12** [comment]
The log output was just pushed.

**Comment by marcjulian** [comment]
@pjlamb12 thanks for this PR. I will merge it tomorrow and update the readme about the breaking changes and the new config option.

**Comment by pjlamb12** [comment]
Sweet! I was just thinking, I wonder if we should add a line in the README about putting the dates in strings for now.

**Comment by pjlamb12** [comment]
Just a heads up, here's a fix in Scully for this issue: https://github.com/scullyio/scully/pull/1140

**Comment by marcjulian** [comment]
@pjlamb12 awesome 💯 ! We can wait for it to be release and than we can remove dayjs and not introduce a breaking change.

**Comment by pjlamb12** [comment]
Although it will still be a breaking change since it's not a tender plugin anymore, right?

**Comment by marcjulian** [comment]
Yes true. I will make that clear anyways and it will require at least `@scullyio/scully@1.0.5`

**Comment by marcjulian** [comment]
@pjlamb12 thanks for your help!! 👍 new PRs are always welcome.

**Comment by pjlamb12** [comment]
No problem, glad to help, and I'm excited to implement this in my own site! Thank you for creating the plugin and finishing up the pr!

**Comment by marcjulian** [comment]
Your welcome. Release is published! I bumped it to 1.0.0 as it is a new plugin type. If you find anything missing in the readme or anywhere let me know!

**Pull request merged**